components/formats-api/src/loci/formats/MetadataList.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * #%L
+ * primary reader and writer APIs
+ * %%
+ * Copyright (C) 2018 Open Microscopy Environment:
+ *   - Board of Regents of the University of Wisconsin-Madison
+ *   - Glencoe Software, Inc.
+ *   - University of Dundee
+ * %%
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package loci.formats;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+
+/**
+ * A container of metadata for FormatReader and other reader implementations.  The structure is designed to contain
+ * data for individual image series', including sub-resolutions.  This is implemented internally as a list of
+ * lists indexed by series and resolution index.  However, it may be used to store any data in a two-level
+ * structure.
+ *
+ * The container may be sized up front at the primary or at both levels.  Or either level may be appended to
+ * individually.  Both patterns are used by many different readers.
+ *
+ * This class was originally developed to store the CoreMetadata for image series and sub-resolutions.  However, it
+ * may be used to store any other data, for example reader-specific data including IFD mappings, extended
+ * metadata, etc.
+ *
+ * @param <T> The type of metadata to store.
+ */
+public class MetadataList<T> {
+  protected ArrayList<ArrayList<T>> data = new ArrayList<>();
```

Comment:
Is it okay to declare `data` as being simply `List<List<T>>`?

Reply:
Annoyingly, it doesn't work.  Nested generics cause javac to error out.  I tried and failed, and also worked through several ways to do it with Riad, without success.

Reply:
I'll fetch your branch and take a look.

Reply:
```diff
diff --git a/components/formats-api/src/loci/formats/CoreMetadataList.java b/components/formats-api/src/loci/formats/CoreMetadataList.java
index b0b84ab076..6b8d627f24 100644
--- a/components/formats-api/src/loci/formats/CoreMetadataList.java
+++ b/components/formats-api/src/loci/formats/CoreMetadataList.java
@@ -105,7 +105,7 @@ public class CoreMetadataList extends MetadataList<CoreMetadata> {
   }
 
   public void reorder() {
-    for (ArrayList<CoreMetadata> s : data) {
+    for (List<CoreMetadata> s : data) {
       Collections.sort(s, comparator);
     }
   }
diff --git a/components/formats-api/src/loci/formats/MetadataList.java b/components/formats-api/src/loci/formats/MetadataList.java
index 5fa7215b50..b1a09f9ac3 100644
--- a/components/formats-api/src/loci/formats/MetadataList.java
+++ b/components/formats-api/src/loci/formats/MetadataList.java
@@ -52,7 +52,7 @@ import java.util.ArrayList;
  * @param <T> The type of metadata to store.
  */
 public class MetadataList<T> {
-  protected ArrayList<ArrayList<T>> data = new ArrayList<>();
+  protected List<List<T>> data = new ArrayList<>();
 
   // Construct an empty list.
   public MetadataList() {
```
Does this not suffice? It builds cleanly for me.